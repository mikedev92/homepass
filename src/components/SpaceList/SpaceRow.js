// @flow
import React from 'react';
import { Row, Left, View, Text, ListItem, Body, Thumbnail } from 'native-base';
import { withNavigation } from 'react-navigation';
import gql from 'graphql-tag';
import moment from 'moment';
import { compose, setStatic } from 'recompose';
import { get } from 'lodash';

type Props = {
	space: Space,
	navigation: Object,
	last: Boolean
};
function SpaceRow({ space, navigation, last }: Props) {
	const { id, title, subtitle, imageUrl, nextEvent } = space;
	const memberEdges: SpaceMemberEdge[] = get(space, 'members.edges') || [];
	return (
		<ListItem thumbnail onPress={() => navigation.navigate('Space', { id, title })} last={last}>
			<Left>
				<Thumbnail square style={{ borderRadius: 5 }} source={{ uri: imageUrl }} />
			</Left>
			<Body>
				<Text>
					{title}
					{'\n'}
					<Text note>
						{subtitle}
					</Text>
					{'\n'}
					{nextEvent &&
						<Text note>
							{nextEvent.name} {moment(nextEvent.startDate).fromNow()}
						</Text>}
				</Text>
				<Row>
					<View style={{ flex: 1 }} />
					{memberEdges.map(member => member && member.node && member.node.user && member.node.user.imageUrl && <Thumbnail xSmall key={member.node.id} source={{ uri: member.node.user.imageUrl }} style={{ marginHorizontal: 2 }} />)}
				</Row>
			</Body>
		</ListItem>
	);
}

export default compose(
	setStatic('fragments', {
		space: gql`
			fragment SpaceRowSpace on Space {
				id
				title
				imageUrl
				subtitle
				nextEvent {
					startDate
					endDate
					name
				}
				members {
					edges {
						node {
							id
							user {
								id
								imageUrl
							}
						}
					}
				}
			}
		`
	}),
	withNavigation
)(SpaceRow);
