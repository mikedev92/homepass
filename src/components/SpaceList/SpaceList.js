// @flow
import React from 'react';
import { SectionList } from 'react-native';
import { Header, Left, Button, Icon, Body, Right, Text, Container, Separator } from 'native-base';
import { graphql } from 'react-apollo';
import gql from 'graphql-tag';
import { compose, withState } from 'recompose';
import { get, lowerCase, upperCase, groupBy } from 'lodash';
import Search from 'react-native-search-box';
import { connect } from 'react-redux';
import SpaceRow from './SpaceRow';
import { graphqlLoader } from '../common';
import { spaceSelector } from '../../reducers';

type Props = {
	data: ApolloData<void>,
	filters: any,
	searchText: string,
	setSearchText: string => void,
	navigation: Navigation
};

function filterSpaces(spaces?: Array<SpaceEdge>, searchText: string, filters): Array<SpaceEdge> {
	if (!spaces) {
		return [];
	}
	const shouldFilterByTags = Object.keys(filters).reduce((prevState, key) => filters[key] || prevState, false);
	return spaces.filter(space => {
		const title = get(space, 'node.title', '');
		const tags = get(space, 'node.tags', []);
		return lowerCase(title).includes(lowerCase(searchText)) && (!shouldFilterByTags || tags.find(tag => filters[tag]));
	});
}

function generateSectionForSpaces(key: string, groupedSpaces: Object) {
	const data = groupedSpaces[key] || [];
	return {
		data,
		key
	};
}
function generateSectionsForSpaces(spaces: Array<SpaceEdge>) {
	const groupedSpaces = groupBy(spaces, spaceEdge => (get(spaceEdge, 'node.viewerMember') ? 'my places' : 'other places'));
	return [generateSectionForSpaces('my places', groupedSpaces), generateSectionForSpaces('other places', groupedSpaces)];
}

function Spaces({ setSearchText, data, filters, searchText, navigation }: Props) {
	const filteredSpaces = filterSpaces(get(data, 'viewer.spaces.edges'), searchText, filters);
	const sections = generateSectionsForSpaces(filteredSpaces);
	return (
		<Container>
			<Header>
				<Left>
					<Button onPress={() => navigation.goBack()} transparent primary>
						<Icon name="arrow-back" />
					</Button>
				</Left>
				<Body />
				<Right>
					<Button onPress={() => navigation.navigate('SpaceFilters')} transparent primary>
						<Icon name="options" />
					</Button>
				</Right>
			</Header>
			<Search backgroundColor="transparent" onChangeText={setSearchText} titleCancelColor="black" />
			<SectionList
				renderSectionHeader={({ section: { key } }) =>
					<Separator bordered>
						<Text note>
							{upperCase(key)}
						</Text>
					</Separator>}
				renderItem={({ item }: { item: { node: Space }, index: number }) => <SpaceRow space={item.node} last />}
				sections={sections}
				onRefresh={data.refetch}
				refreshing={data.networkStatus === 4}
				keyExtractor={({ node: { id } }) => id}
			/>
		</Container>
	);
}

const SpaceQuery = gql`
  query SpaceQuery {
      viewer {
        userId
        spaces {
          edges {
            node {
              id
              title
              viewerMember {
                id
              }
              tags
              ... SpaceRowSpace
            }
          }
        }
      }
  }
  ${SpaceRow.fragments.space}
`;

export default compose(graphql(SpaceQuery), graphqlLoader, connect(spaceSelector), withState('searchText', 'setSearchText', ''))(Spaces);
