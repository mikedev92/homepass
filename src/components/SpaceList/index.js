// @flow
import SpaceList from './SpaceList';

export default SpaceList;
