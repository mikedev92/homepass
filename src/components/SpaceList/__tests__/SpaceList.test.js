import React from 'react';
import { shallow } from 'enzyme';
import SpaceList from '../SpaceList';

describe('<SpaceList />', () => {
  it('renders a segmented list of properties', () => {
    const wrapper = shallow(
      <SpaceList
        filters={{}}
        searchText=""
        data={{
          viewer: {
            spaces: {
              edges: [{
                node: {
                  title: 'a',
                  isViewerMember: true,
                },
              }, {
                node: {
                  title: 'b',
                  isViewerMember: false,
                },
              }],
            },
          },
        }}
      />,
    );
    expect(wrapper).toMatchSnapshot();
  });
});

