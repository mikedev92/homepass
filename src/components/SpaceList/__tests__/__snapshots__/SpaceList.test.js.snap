// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<SpaceList /> renders a segmented list of properties 1`] = `
<Styled(Container)>
  <Styled(Header)>
    <Styled(Left)>
      <Styled(Button)
        onPress={[Function]}
        primary={true}
        transparent={true}
      >
        <Styled(Icon)
          name="arrow-back"
        />
      </Styled(Button)>
    </Styled(Left)>
    <Styled(Body) />
    <Styled(Right)>
      <Styled(Button)
        onPress={[Function]}
        primary={true}
        transparent={true}
      >
        <Styled(Icon)
          name="options"
        />
      </Styled(Button)>
    </Styled(Right)>
  </Styled(Header)>
  <Search
    backgroundColor="transparent"
    blurOnSubmit={false}
    editable={true}
    keyboardShouldPersist={false}
    onChangeText={[Function]}
    placeholderCollapsedMargin={15}
    placeholderExpandedMargin={20}
    searchIconCollapsedMargin={25}
    searchIconExpandedMargin={10}
    shadowColor="#000"
    shadowOffsetHeightCollapsed={2}
    shadowOffsetHeightExpanded={4}
    shadowOffsetWidth={0}
    shadowOpacityCollapsed={0.12}
    shadowOpacityExpanded={0.24}
    shadowRadius={4}
    shadowVisible={false}
    titleCancelColor="black"
  />
  <SectionList
    data={Array []}
    disableVirtualization={false}
    getItem={[Function]}
    getItemCount={[Function]}
    horizontal={false}
    initialNumToRender={10}
    keyExtractor={[Function]}
    maxToRenderPerBatch={10}
    onEndReached={[Function]}
    onEndReachedThreshold={2}
    refreshing={false}
    removeClippedSubviews={true}
    renderItem={[Function]}
    renderScrollComponent={[Function]}
    renderSectionHeader={[Function]}
    sections={
      Array [
        Object {
          "data": Array [],
          "key": "my places",
        },
        Object {
          "data": Array [
            Object {
              "node": Object {
                "isViewerMember": true,
                "title": "a",
              },
            },
            Object {
              "node": Object {
                "isViewerMember": false,
                "title": "b",
              },
            },
          ],
          "key": "other places",
        },
      ]
    }
    stickySectionHeadersEnabled={true}
    updateCellsBatchingPeriod={50}
    windowSize={21}
  />
</Styled(Container)>
`;
